### PR TITLE
Join all connect threads while closing connection

### LINKED ISSUE #300: Threads are not stopped on drop/close()
Make sure that these boxes are checked before submitting your issue -- thank you!

 - [x] Included below version and environment information
 - [x] Included a [Minimal, Complete, and Verifiable example] (https://stackoverflow.com/help/mcve)

#### NATS version (`grep 'name = "nats"' Cargo.lock -A 1`)

0.16.0

#### rustc version (`rustc --version` - we support Rust 1.41 and up)

rustc 1.57.0 (f1edd0429 2021-11-29)

#### OS/Container environment:

Ubuntu 20.04.3 LTS

#### Steps or code to reproduce the issue:

Available at https://github.com/ShellWowza/nats-testcase

#### Expected result:

When the Connection is either dropped or closed, all NATS threads are joined on before the return from close() or the Drop implementation, so that the slog logger held by the reconnect/disconnect callback is dropped and therefore flushes at the end of main().

#### Actual result:

NATS threads are not stopped immediately on close() or Drop, resulting in a thread holding onto the logger via a reconnect/disconnect callback past the point where main() returns, which means it does not flush - not all 200-odd lines buffered in the logger are printed.

### PR DESCRIPTION
**Pull request by Jarema**
closes #300

### CODE REVIEW

File: src/client.rs [hunk 1]
```diff
@@ -142,6 +144,7 @@ impl Client {
             server_info: Arc::new(Mutex::new(ServerInfo::default())),
             shutdown: Arc::new(Mutex::new(false)),
             options: Arc::new(options),
+            join_handles: Arc::new(Mutex::new(Vec::new())),
```
Comment (caspervonb) [inline src/client.rs:147]: Sure we don't want to just name these? should be two handles iirc (looked into this on Friday).

Comment (Jarema) [reply src/client.rs]: agree, changed.

**Comment by Jarema** [comment]
Drop on `Inner` starts shutdown, but shutdown itself does not wait until it's finished, just triggers it.

Do we want to change it?

**Comment by caspervonb** [comment]
> Drop on Inner starts shutdown, but shutdown itself does not wait until it's finished, just triggers it.

Already clarified offline but for completeness yes, let's call close.

**Comment by Jarema** [comment]
@caspervonb Added break in flush thread, so closing client on dropped server won't hang.

**Review by caspervonb — APPROVED** [review]

**Pull request merged**